compiler/runtime/src/main/kotlin/runtime/driver.kt

```diff
@@ -0,0 +1,195 @@
+    override fun getCell(vararg relPos: Int): Int {
+        val overEdge = relPos.mapIndexed { index, p ->
+            if (dimTypes[index] != DimensionType.EDGE) {
+                (pos[index] + p) >= dims[index] || (pos[index] + p) < 0
+            } else {
+                false
+            }
+        }.contains(true)
+        return if (overEdge) {
+            edge
+        } else {
+            world.getCell(*relPos.mapIndexed { index, p ->
+                val x = (pos[index] + p).rem(dims[index])
+                // We need a bit of logic to convert remainder operator to modulus operator,
+                // otherwise we are gonna have problems with negative numbers
+                if (x < 0) {
+                    x + dims[index]
+                } else {
+                    x
+                }
+            }.toIntArray())
+        }
+    }
+}
+
+interface IWorldType {
+    fun getWorldView(world: World, dims: List<Int>, pos: List<Int>): IWorldView
+}
+
+@Deprecated("Use MultiWorldType")
+class WrappingWorldType: IWorldType {
+    override fun getWorldView(world: World, dims: List<Int>, pos: List<Int>): WrappingWorldView {
+        return WrappingWorldView(world, dims, pos)
+    }
+}
+
+@Deprecated("Use MultiWorldType")
+class EdgeWorldType(private val edge: Int): IWorldType {
+    override fun getWorldView(world: World, dims: List<Int>, pos: List<Int>): IWorldView {
+        return EdgeWorldView(world, dims, pos, edge)
+    }
+}
+
+class MultiWorldType(private val edge: Int, private val dimTypes: List<DimensionType>): IWorldType {
+    override fun getWorldView(world: World, dims: List<Int>, pos: List<Int>): IWorldView {
+        return MultiWorldView(world, dims, pos, edge, dimTypes)
+    }
+}
+
+open class Driver(private val worldConfig: WorldConfiguration, private val program: IProgram, private val worldType: IWorldType) {
+    var worldCurrent = World(worldConfig.dims)
+    var worldNext = World(worldConfig.dims)
+
+    open fun run() {
+        while(true) {
+            update()
+        }
+    }
+
+    open fun update() {
+        fun iterate(subDims: List<Int>, pos: List<Int> = listOf()) {
+            if (subDims.size == 1) {
+                var x = subDims[0] - 1
+                while (x >= 0) {
+                    val currentPos = pos + listOf(x)
+                    worldNext.setCell(
+                        *currentPos.toIntArray(),
+                        state = program.updateCell(
+                            worldView = worldType.getWorldView(worldCurrent, worldConfig.dims, currentPos)
+                        )
+                    )
+                    x--
+                }
+            } else {
+                var x = subDims[0] - 1
+                while (x >= 0) {
+                    @Suppress("UNCHECKED_CAST")
+                    iterate(subDims.subList(1, subDims.size), pos + listOf(x))
+                    x--
```

Comment:
`@Suppress("UNCHECKED_CAST")` is unnecessary.
`pos + listOf(x)` can be `pos + x` instead
... and if we want to be real fancy `pos + x--` and remove the following `x--`